src/planner/operator/schema.cpp

```diff
@@ -99,6 +99,43 @@ void Schema::clear() {
     clearExpressionsInScope();
 }
 
+size_t Schema::getNumGroups(bool isFlat) const {
+    auto result = 0u;
+    for (auto groupPos : getGroupsPosInScope()) {
+        result += groups[groupPos]->isFlat() == isFlat;
+    }
+    return result;
+}
+
+std::size_t SchemaHasher::operator()(const Schema* const& schema) const {
+    return std::hash<size_t>{}(schema->getNumFlatGroups()) ^
+           std::hash<size_t>{}(schema->getNumUnFlatGroups());
+}
+
+// We use this equality in join order enumeration to make sure at each DP level, we don't just keep
+// the best plan, but keep best plan for each unique factorization schema.
+// In order to balance enumeration time, we use an approximate equality check to reduce computation.
```

Comment:
- Should we also check the equality of num of expressions in the left and right schemas?
- Can you add comments on the approximate equal logic?
```
Two schemas (left and right) are considered approximately equal, if:
1) The left schema and the right schema have same number of flat and unflat groups.
2) Any expression in the left schema also exists in the right schema, and its corresponding factorization groups in the left and right schemas are both flat or both unflat.
```

Reply:
done